compiler/rustc_mir/src/const_eval/machine.rs

```diff
@@ -17,60 +15,13 @@ use rustc_span::symbol::{sym, Symbol};
 use rustc_target::abi::{Align, Size};
 
 use crate::interpret::{
-    self, compile_time_machine, AllocId, Allocation, Frame, GlobalId, ImmTy, InterpCx,
-    InterpResult, Memory, OpTy, PlaceTy, Pointer, Scalar,
+    self, compile_time_machine, AllocId, Allocation, Frame, ImmTy, InterpCx, InterpResult, Memory,
+    OpTy, PlaceTy, Pointer, Scalar,
 };
 
 use super::error::*;
 
 impl<'mir, 'tcx> InterpCx<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>> {
-    /// Evaluate a const function where all arguments (if any) are zero-sized types.
-    /// The evaluation is memoized thanks to the query system.
-    ///
-    /// Returns `true` if the call has been evaluated.
-    fn try_eval_const_fn_call(
-        &mut self,
-        instance: ty::Instance<'tcx>,
-        ret: Option<(PlaceTy<'tcx>, mir::BasicBlock)>,
-        args: &[OpTy<'tcx>],
-    ) -> InterpResult<'tcx, bool> {
-        trace!("try_eval_const_fn_call: {:?}", instance);
-        // Because `#[track_caller]` adds an implicit non-ZST argument, we also cannot
-        // perform this optimization on items tagged with it.
-        if instance.def.requires_caller_location(self.tcx()) {
-            return Ok(false);
-        }
-        // Only memoize instrinsics. This was added in #79594 while adding the `const_allocate` intrinsic.
-        // We only memoize intrinsics because it would be unsound to memoize functions
-        // which might interact with the heap.
-        // Additionally, const_allocate intrinsic is impure and thus should not be memoized;
-        // it will not be memoized because it has non-ZST args
-        if !matches!(instance.def, InstanceDef::Intrinsic(_)) {
-            return Ok(false);
-        }
-        // For the moment we only do this for functions which take no arguments
-        // (or all arguments are ZSTs) so that we don't memoize too much.
-        if args.iter().any(|a| !a.layout.is_zst()) {
-            return Ok(false);
-        }
-
-        let dest = match ret {
-            Some((dest, _)) => dest,
-            // Don't memoize diverging function calls.
-            None => return Ok(false),
-        };
-
-        let gid = GlobalId { instance, promoted: None };
-
-        let place = self.eval_to_allocation(gid)?;
```

Comment:
@oli-obk is there any further cleanup possible due to this? Looks like The const-eval queries support passing in the `Instance` of an argumentless const-fn, and then return the result. I feel like taht would need some special handling *somewhere* and I think we should stop doing that.

Reply:
Judging from [the PR that added this](https://github.com/rust-lang/rust/pull/66294/files), 
* `eval_body_using_ecx` had an assertion `assert!(body.arg_count == 0);`. We should get that back... but also it's not enough! Even argumentless functions can do allocations and thus should not be allowed as the subject of a query.
* The doc comment of `GlobalId` needs updating to no longer mention "A const fn where all arguments (if any) are zero-sized types".

Reply:
>     * `eval_body_using_ecx` had an assertion `assert!(body.arg_count == 0);`. We should get that back... but also it's not enough! Even argumentless functions can do allocations and thus should not be allowed as the subject of a query.

I think the right way to specify that assertion is

```rust
assert!(promoted.is_some() || matches!(tcx.hir().body_const_context(def_id), Some(ConstContext::Const | ConstContext::Static(_))));
```

Which makes sure that we really only allow promoteds, constants or statics.

Reply:
```rust
assert!(cid.promoted.is_some() || matches!(ecx.tcx.hir().body_const_context(def_id), Some(ConstContext::Const | ConstContext::Static(_))));
```
like this in the `eval_body_using_ecx` right after first debug output ?

Reply:
Yes, that is the right place.